File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(a: list[str], p: Callable[[str], bool]) -> str:
    for s in a:
        if p(s):
            return s.lower()
    return None


# Write your powers generator here

def powers_generator(*, base, limit):
    number = 1
    while number <= limit:
        yield number
        number *= base
        
# Write your say function here
def say(word=None):
    # Base case: if called without an argument, return an empty string
    if word is None:
        return ""
    
    # Recursive function to accumulate words
    def inner_say(next_word=None, words=[word]):
        if next_word is None:
            return ' '.join(words)
        else:
            words.append(next_word)
            return lambda w=None: inner_say(w, words)
    
    return inner_say

# Write your line count function here
def meaningful_line_count(file_path: str):
        count = 0
        try:
            with open(file_path, "r") as file:
                for line in file:
                    stripped_line = line.strip()
                    # Check if the line is not empty and does not start with '#'
                    if stripped_line and not stripped_line.startswith('#'):
                        count += 1
        except FileNotFoundError:
            raise FileNotFoundError("No such file")
        
        return count
# Write your Quaternion class here
class Quaternion:
    def __init__(self, a: float, b: float, c: float, d: float) -> None:
        self.a = a
        self.b = b
        self.c = c
        self.d = d
    
    def __add__(self, other):
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
    
    def __mul__(self, other):
        if isinstance(other, Quaternion):
            # Quaternion multiplication formula
            a = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
            b = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
            c = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
            d = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
            return Quaternion(a, b, c, d)
        elif isinstance(other, (int, float)):
            # Scalar multiplication
            return Quaternion(self.a * other, self.b * other, self.c * other, self.d * other)
        else:
            raise TypeError("Multiplication only supports Quaternion or scalar types.")
    
    @property
    def conjugate(self):
        """Return the conjugate of the quaternion."""
        return Quaternion(self.a, -self.b, -self.c, -self.d)
    
    def __eq__(self, other):
        """Equality check for quaternions."""
        return (self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d)
    
    @property
    def coefficients(self):
        """Return the coefficients (a, b, c, d) as a tuple."""
        return (self.a, self.b, self.c, self.d)
    
    def __repr__(self):
        parts = []
        
        # Scalar part (a)
        if self.a != 0:
            parts.append(f"{self.a}")

        # Imaginary parts (b, c, d)
        if self.b != 0:
            if self.b == 1:
                parts.append("+i")
            elif self.b == -1:
                parts.append("-i")
            else:
                parts.append(f"{self.b:+}i")
        
        if self.c != 0:
            if self.c == 1:
                parts.append("+j")
            elif self.c == -1:
                parts.append("-j")
            else:
                parts.append(f"{self.c:+}j")

        if self.d != 0:
            if self.d == 1:
                parts.append("+k")
            elif self.d == -1:
                parts.append("-k")
            else:
                parts.append(f"{self.d:+}k")
        
        # Join the parts together and remove any leading "+" signs
        result = "".join(parts)
        return result.lstrip('+') if result else '0'

File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(a, p)
  for _, str in ipairs(a) do
    if p(str) then
      return string.lower(str)
    end
  end
  return nil
end

-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.create(function()
    local power = 0
    local result = 1
    while result <= limit do
      coroutine.yield(result)
      power = power + 1
      result = base^power
    end
  end)
end

-- Write your say function here
function say(word)
  local words = {}

  local function addWord(str)
  if str then
    table.insert(words, str)
    return addWord
  else
    return table.concat(words, " ")
  end
  end

  if word then
    return addWord(word)
  else
    return ""
  end
end

-- Write your line count function here
function meaningful_line_count(filename)
    local file = io.open(filename, "r")
    if not file then
        return error("No such file")
    end

    local count = 0
    for line in file:lines() do
        local trimmed_line = line:match("^%s*(.-)%s*$") -- trim whitespace from the line
        if trimmed_line ~= "" and trimmed_line:sub(1, 1) ~= "#" then
            count = count + 1
        end
    end

    file:close()
    return count
end

-- Write your Quaternion table here
Quaternion = {}
Quaternion.__index = Quaternion
function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a -- real
  self.b = b -- i
  self.c = c -- j
  self.d = d -- k
  return self
end
function Quaternion.__add(a,b)
  return Quaternion.new(a.a + b.a, a.b + b.b, a.c + b.c, a.d + b.d)
end
function Quaternion.__mul(x,y)
  return Quaternion.new(
    x.a * y.a - x.b * y.b - x.c * y.c - x.d * y.d,
    x.a * y.b + x.b * y.a + x.c * y.d - x.d * y.c,
    x.a * y.c - x.b * y.d + x.c * y.a + x.d * y.b,
    x.a * y.d + x.b * y.c - x.c * y.b + x.d * y.a)
end
function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end
function Quaternion.__eq(a,b)
  return (a.a == b.a and a.b == b.b) and (a.c == b.c and a.d == b.d)
end
function Quaternion:conjugate()
  return Quaternion.new(self.a, - self.b, - self.c, - self.d)
end
function Quaternion:__tostring()
  local output = ""
  
  for i, element in ipairs(self:coefficients()) do
    
    if element ~= 0 then
    
      if element < 0 then
        output = output .. "-"
      else
        if output ~= "" then
          output = output .. "+"
        end
      end
      
      local abs = math.abs(element)
      
      if i == 1 then
        output = output .. tostring(abs)
      else
        if abs ~= 1 then
          output = output .. tostring(abs)
        end
      
        if i == 2 then
          output = output .. "i"
        else
          if i == 3 then
            output = output .. "j"
          else
            output = output .. "k"
          end
        end
      end
    
    end
  end
  
  if output == "" then
    return "0"
  else
    return output
  end
end
